language/move-vm/runtime/src/data_cache.rs

```diff
@@ -72,9 +72,12 @@ impl<'r, 'l, S: MoveResolver> TransactionDataCache<'r, 'l, S> {
     /// published modules.
     ///
     /// Gives all proper guarantees on lifetime of global data as well.
-    pub(crate) fn into_effects(self) -> PartialVMResult<(ChangeSet, Vec<Event>)> {
+    pub(crate) fn into_effects(mut self) -> (PartialVMResult<(ChangeSet, Vec<Event>)>, S) {
+        (self.impl_into_effects(), self.remote)
+    }
+    fn impl_into_effects(&mut self) -> PartialVMResult<(ChangeSet, Vec<Event>)> {
         let mut change_set = ChangeSet::new();
-        for (addr, account_data_cache) in self.account_map.into_iter() {
+        for (addr, account_data_cache) in std::mem::take(&mut self.account_map).into_iter() {
```

Comment:
Not sure if I understand the need for splitting this into two functions. It seems like taking the data cache by value and then destructuring it into its parts would still work fine here and remove the need for the `std::mem::take`s below.